### PR TITLE
Section refs finder tweaks

### PR DESCRIPTION
**Pull request by goose-life**
deals with the following real-life examples encountered since deploying:

(1) A person who contravenes sections [4](#section-4)(1) and (2), 6(3), 10(1) and (2), 11(1), 12(1), 19(1), 19(3), 20(1), 20(2), 21(1), 22(1), 24(1), 25(3), (4) , (5) and (6) , 26(1), (2), (3) and (5), 28(1), (2) and (3) is guilty of an offence.
(https://edit.laws.africa/documents/3376/)

in sub-[section 1](#section-1) will

sections [64](#section-64), [65](#section-65), and 67 of this Act

under [sections 35](#section-35),36, or 37 of this Act

### CODE REVIEW

File: indigo/analysis/refs/base.py [hunk 1]
```diff
@@ -196,13 +197,21 @@ class SectionRefsFinderENG(BaseInternalRefsFinder):
         r'''\b
         (
           (?P<ref>
-            sections?\s+
+            (?<!-)sections?\s+
```
Comment (longhotsummer) [inline indigo/analysis/refs/base.py]: First time I've seen a negative look-behind assertion!

File: indigo/analysis/refs/base.py [hunk 2]
```diff
@@ -196,13 +197,21 @@ class SectionRefsFinderENG(BaseInternalRefsFinder):
...
             (?P<num>\d+[A-Z0-9]*)  # first section number, including subsections
           )
-          (\s*\([A-Z0-9]+\))*      # bracketed subsections of first number
+          (
+            (\s*(,|(,\s*)?and|(,\s*)?or)\s*)*          # list separators
```
Comment (longhotsummer) [inline indigo/analysis/refs/base.py]: How about:

`(\s*(,|and|or))*`

The final `*` allows multiple repetitions, which handles `, and` and `, or` so they don't need to be special cased.

Comment (longhotsummer) [reply indigo/analysis/refs/base.py]: ditto for the separators below, too.

File: indigo/analysis/refs/base.py [hunk 3]
```diff
@@ -212,7 +221,7 @@ class SectionRefsFinderENG(BaseInternalRefsFinder):
     # individual numbers in the list grouping above
     # we use <ref> and <num> named captures so that the is_valid and make_ref
     # methods can handle matches from both ref_re and this re.
-    item_re = re.compile(r'(?P<ref>(?P<num>\d+[A-Z0-9]*))(\s*\([A-Z0-9]+\))*', re.IGNORECASE)
+    item_re = re.compile(r'(?P<ref>(?P<num>(?<!\()\d+[A-Z0-9]*(?!\))))(\s*\([A-Z0-9]+\))*', re.IGNORECASE)
```
Comment (longhotsummer) [inline indigo/analysis/refs/base.py]: Can you add a comment example of what this guards against?

File: indigo/analysis/refs/base.py (hunk 2 above)
Comment (goose-life) [reply indigo/analysis/refs/base.py]: thanks, yours is much better!

File: indigo/analysis/refs/base.py (hunk 3 above)
Comment (goose-life) [reply indigo/analysis/refs/base.py]: done, shout if it's still not clear!

**Pull request merged**